src/test/checkqueue_tests.cpp

```diff
@@ -0,0 +1,402 @@
+        while (total) {
+            size_t r = GetRand(10);
+            std::vector<UniqueCheck> vChecks;
+            for (size_t k = 0; k < r && total; k++)
+                vChecks.emplace_back(--total);
+            control.Add(vChecks);
+        }
+    }
+    bool r = true;
+    for (size_t i = 0; i < COUNT; ++i)
+        r = r && UniqueCheck::results.count(i) == 1;
+    BOOST_REQUIRE(r);
+    tg.interrupt_all();
+    tg.join_all();
+}
+
+
+// Test that blocks which might allocate lots of memory free their memory agressively.
+//
+// This test attempts to catch a pathalogical case where by lazily freeing
+// checks might mean leaving a check un-swapped out, and decreasing by 1 each
+// time could leave the data hanging across a sequence of blocks.
+//
+// This test (failing) is dependent on not being able to handle
+// 1000*200 MB of memory, and able to handle 600MB of memory
+BOOST_AUTO_TEST_CASE(test_CheckQueue_Memory)
+{
+    auto queue = std::shared_ptr<Memory_Queue>(new Memory_Queue {128});
+    boost::thread_group tg;
+    for (auto x = 0; x < nScriptCheckThreads; ++x) {
+       tg.create_thread([=]{queue->Thread();});
+    }
+    for (size_t i = 9999; i < 9999; --i) {
+        size_t total = i;
+        {
+            CCheckQueueControl<MemoryCheck> control(queue.get());
+            while (total) {
+                size_t r = GetRand(10);
+                std::vector<MemoryCheck> vChecks;
+                for (size_t k = 0; k < r && total; k++) {
+                    total--;
+                    // Each iteration leaves data at the front, back, and middle
+                    // to catch any sort of deallocation failure
+                    vChecks.emplace_back(total == 0 || total == i || total == i/2);
+                }
+                control.Add(vChecks);
+            }
+        }
+    }
+    tg.interrupt_all();
+    tg.join_all();
+}
+
+// Test that a new verification cannot occur until all checks 
+// have been destructed
+BOOST_AUTO_TEST_CASE(test_CheckQueue_FrozenCleanup)
+{
+    auto queue = std::shared_ptr<FrozenCleanup_Queue>(new FrozenCleanup_Queue {128});
+    boost::thread_group tg;
+    for (auto x = 0; x < nScriptCheckThreads; ++x) {
+       tg.create_thread([=]{queue->Thread();});
+    }
+    std::atomic<bool> finished{false};
+    std::thread t0([&]() {
+        CCheckQueueControl<FrozenCleanupCheck> control(queue.get());
+        FrozenCleanupCheck::frozen = true;
+        std::vector<FrozenCleanupCheck> vChecks;
+        vChecks.emplace_back();
+        control.Add(vChecks);
+        control.Wait(); // Hangs here
+        finished = true;
+    });
+    // We should never read a finished status
+    //
+    // Try waiting a second (which should be plenty of time to reach the hang
+    // point in t0)
+    //
+    // Note that this cannot cause a spurious failure, only could mean that
+    // the test doesn't actually end up checking that control waited.
+    MilliSleep(1000);
```

Comment:
I think you could easily make this test deterministic as well, eliminating the long sleep and the while (frozen) busy loops. Would just need to have ~FrozenCleanupCheck increment a counter and signal a conditional variable so you could wait here for enough jobs to be frozen, and then do the boost check. Then this could signal another condition variable to unfreeze the jobs. I think it would be worth changing this to make the test more efficient and reliable.